state/azure/cosmosdb/cosmosdb.go

```diff
@@ -456,23 +419,105 @@ func (c *StateStore) Query(req *state.QueryRequest) (*state.QueryResponse, error
 }
 
 func (c *StateStore) Ping() error {
-	m := c.metadata
+	_, err := c.findCollection()
+	return err
+}
 
-	colls, err := c.client.QueryCollections(c.db.Self, &documentdb.Query{
-		Query: "SELECT * FROM ROOT r WHERE r.id=@id",
-		Parameters: []documentdb.Parameter{
-			{Name: "@id", Value: m.Collection},
-		},
-	})
-	if err != nil {
+// getCollectionLink returns the link to the collection.
+func (c *StateStore) getCollectionLink() string {
+	return fmt.Sprintf("dbs/%s/colls/%s/", c.metadata.Database, c.metadata.Collection)
+}
+
+// getSprocLink returns the link to the stored procedures.
+func (c *StateStore) getSprocLink(sprocName string) string {
+	return fmt.Sprintf("dbs/%s/colls/%s/sprocs/%s", c.metadata.Database, c.metadata.Collection, sprocName)
+}
+
+func (c *StateStore) checkStoredProcedures() error {
+	var ver int
+	err := c.client.ExecuteStoredProcedure(c.getSprocLink(versionSpName), nil, &ver, documentdb.PartitionKey("1"))
+	if err == nil {
+		c.logger.Debugf("Cosmos DB stored procedure version: %d", ver)
+	}
+	if err != nil || (err == nil && ver != spVersion) {
+		return fmt.Errorf("Dapr requires stored procedures created in Cosmos DB before it can be used as state store. Those stored procedures are currently not existing or are using a different version than expected. When you authenticate using Azure AD we cannot automatically create them for you: please start this state store with a Cosmos DB master key just once so we can create the stored procedures for you; otherwise, you can check our docs to learn how to create them yourself: https://aka.ms/dapr/cosmosdb-aad") // nolint:stylecheck
+	}
+	return nil
+}
+
+func (c *StateStore) ensureStoredProcedures() error {
+	spLink := c.getSprocLink(storedProcedureName)
+	verSpLink := c.getSprocLink(versionSpName)
+
+	// get a link to the sp's
+	sp, err := c.client.ReadStoredProcedure(spLink)
+	if err != nil && !isNotFoundError(err) {
 		return err
-	} else if len(colls) == 0 {
-		return fmt.Errorf("collection %s for CosmosDB state store not found", m.Collection)
+	}
+	verSp, err := c.client.ReadStoredProcedure(verSpLink)
+	if err != nil && !isNotFoundError(err) {
+		return err
+	}
+
+	// check version
+	replace := false
+	if verSp != nil {
+		var ver int
+		err = c.client.ExecuteStoredProcedure(verSpLink, nil, &ver, documentdb.PartitionKey("1"))
+		if err == nil {
+			c.logger.Debugf("Cosmos DB stored procedure version: %d", ver)
+		}
+		if err != nil || (err == nil && ver != spVersion) {
```

Comment:
Note that this checks the actual version and if it's the same as what we were expecting. If the version is different (lower/higher), it triggers the commands to replace the stored procedures.

This should address all cases, including when customers upgrade Dapr, or even when they downgrade it.